rllib/env/multi_agent_env.py

```diff
@@ -590,71 +600,110 @@ def __init__(
     @override(BaseEnv)
     def poll(
         self,
-    ) -> Tuple[MultiEnvDict, MultiEnvDict, MultiEnvDict, MultiEnvDict, MultiEnvDict]:
-        obs, rewards, dones, infos = {}, {}, {}, {}
+    ) -> Tuple[
+        MultiEnvDict,
+        MultiEnvDict,
+        MultiEnvDict,
+        MultiEnvDict,
+        MultiEnvDict,
+        MultiEnvDict,
+    ]:
+        obs, rewards, terminateds, truncateds, infos = {}, {}, {}, {}, {}
         for i, env_state in enumerate(self.env_states):
-            obs[i], rewards[i], dones[i], infos[i] = env_state.poll()
-        return obs, rewards, dones, infos, {}
+            (
+                obs[i],
+                rewards[i],
+                terminateds[i],
+                truncateds[i],
+                infos[i],
+            ) = env_state.poll()
+        return obs, rewards, terminateds, truncateds, infos, {}
 
     @override(BaseEnv)
     def send_actions(self, action_dict: MultiEnvDict) -> None:
         for env_id, agent_dict in action_dict.items():
-            if env_id in self.dones:
+            if env_id in self.terminateds or env_id in self.truncateds:
                 raise ValueError(
                     f"Env {env_id} is already done and cannot accept new actions"
                 )
             env = self.envs[env_id]
             try:
-                obs, rewards, dones, infos = env.step(agent_dict)
+                obs, rewards, terminateds, truncateds, infos = env.step(agent_dict)
             except Exception as e:
                 if self.restart_failed_sub_environments:
                     logger.exception(e.args[0])
                     self.try_restart(env_id=env_id)
-                    obs, rewards, dones, infos = e, {}, {"__all__": True}, {}
+                    obs = e
```

Comment:
just wondering -- how do these get handled downstream? no need to answer for this pr, just curious

Reply:
The Sampler will return the error as `obs` in the call to `env.poll()`. This way, it knows that one of the sub-envs has crashed and needs to be restarted. All the other pieces of information are discarded (rewards, etc..) and the entire episode as well (not used for training anymore).